compiler/rustc_middle/src/query/plumbing.rs

```diff
@@ -123,13 +123,25 @@ impl<'tcx> TyCtxt<'tcx> {
         TyCtxtEnsureOk { tcx: self }
     }
 
-    /// Returns a transparent wrapper for `TyCtxt`, which ensures queries
-    /// are executed instead of just returning their results.
+    /// Wrapper that calls queries in a special "ensure done" mode, for callers
+    /// that don't need the return value and just want to guarantee that the
+    /// query won't be executed in the future, by executing it now if necessary.
     ///
-    /// This version verifies that the computed result exists in the cache before returning.
+    /// This is useful for queries that read from a [`Steal`] value, to ensure
+    /// that they are executed before the query that will steal the value.
+    ///
+    /// Unlike [`Self::ensure_ok`], a query with all-green inputs will only be
+    /// skipped if its return value is stored in the disk-cache. This is still
+    /// more efficient than a regular query, because in that situation the
+    /// return value doesn't necessarily need to be decoded.
+    ///
+    /// (As with all query calls, execution is also skipped if the query result
+    /// is already cached in memory.)
+    ///
+    /// [`Steal`]: rustc_data_structures::steal::Steal
     #[inline(always)]
-    pub fn ensure_with_value(self) -> TyCtxtEnsureWithValue<'tcx> {
-        TyCtxtEnsureWithValue { tcx: self }
+    pub fn ensure_done(self) -> TyCtxtEnsureDone<'tcx> {
```

Comment:
I guess `ensure_green_and_value_cached` is a mouthfull, but `ensure_done` isn't clearer to me than `ensure_with_value`.

Reply:
Both of these query modes are subtle enough that the name can't hope to fully describe them; only docs (and the implementation) can do that.

My goal here is to focus more on the caller's *intent*.

“Ensure that a usable value exists in a cache somewhere” is a means to an end, whereas (if I understand correctly) what the caller actually wants is a guarantee that the query won't execute in the future. So `ensure_done` is intended to be a more memorable shorthand for that desire.

Reply:
FWIW, "green" means nothing to me in this context.^^ So that name isn't clear at all either.

Reply:
I guess if you're mucking with query optimizations some documentation reading is advisable :laughing: